### DISCUSSION TITLE
what does `drag = 1` even mean?

### DISCUSSION THREAD

**Discussion by dragoncoder047** [kaplayjs/kaplay#424] [Q&A] [answered]
I was tweaking the physics of my platfomer and found myself setting `drag` on the body to numbers greater than 1 to make it stop quickly upon hitting the ground... this doesn't seem right...

Looking in `body.ts` the documentation string says "`velocity *= 1 - drag` every frame":

https://github.com/kaplayjs/kaplay/blob/37c8d67dfb679847ac3a734819aa53460ae5654a/src/components/physics/body.ts#L23-L28

My understanding  is that if the drag is 1, this means that the object will be able to be pushed, but it will not continue to slide after it stops being pushed. This follows from the docs' equation: plug in drag = 1 and you get `velocity *= (1 - 1)` => `velocity *= 0`.

But the code actually does `velocity *= 1 - (drag * dt)`:

https://github.com/kaplayjs/kaplay/blob/37c8d67dfb679847ac3a734819aa53460ae5654a/src/components/physics/body.ts#L408-L409

If drag = 1 is used in the current code's equation, and the frame rate is say, 60fps, then the equation becomes `velocity *= 1 - (1 * 1/60)` => `velocity *= 59/60` which is probably not what was intended.

I tried putting in values larger than 1 for drag, and this does make the object stop faster, but it has caveats: if I use drag = 20, and then the game lags and the frame rate drops below 20fps, then the equation becomes (if the frame rate is currently 12fps) `velocity *= 1 - (20 * 1/12)` => `velocity *= -2/3` and whoops, the object has "magically" reversed direction!

It would maybe be more clear if `drag` had a unit on it.

**Comment by mflerackers** [comment, marked as answer]
Drag is old Kaboom code which still needs to be fixed. It's actually damping, since drag would be through a medium, and depend on geometry. Damping is a ratio, so it should be

this.vel.x *= 1.0 / (1.0 + this.damping * k.dt());
this.vel.y *= 1.0 / (1.0 + this.damping * k.dt());

The physics framerate is always 50fps, regardless of the display framerate.

**Reply by dragoncoder047** [reply]
OK, that makes more sense... so if I want it to stop instantly I just have to use `damping = Infinity`. (that is, once this is fixed).

**Comment by mflerackers** [comment]
Well, yes, but usually damping between 0 and 0.1 is used. If you want
something to stop instantly, just reset the velocity.

On Fri, 4 Oct 2024 at 08:10, dragoncoder047 ***@***.***>
wrote:

> OK, that makes more sense... so if I want it to stop instantly I just have
> to use damping = Infinity. (that is, once this is fixed).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kaplayjs/kaplay/discussions/424#discussioncomment-10838184>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAHIYWZ6VOOTJ4WEYF2IXKLZZXFGLAVCNFSM6AAAAABPJ557J6VHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTAOBTHAYTQNA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Kind Regards,

Marc Flerackers
Software Engineer

**Reply by dragoncoder047** [reply]
But no, think about it if damping=Infinity:

`this.vel *= 1 / (1 + Infinity * 1/50)` 
`this.vel *= 1 / (1 + Infinity)` 
`this.vel *= 1 / Infinity`  
`this.vel *= 0`  
which does the same thing!

**Comment by mflerackers** [comment]
I know, but I don't think damping is something you change in order to stop an object from moving at certain moments in time. And if it is always at infinity, the object won't move when forces are used on it.

**Reply by dragoncoder047** [reply]
I guess I'm looking for friction then...

**Comment by mflerackers** [comment]
Yes, I'm adding friction and restitution to the area component (to be
compatible with other physics engines) once I have time.

On Fri, 4 Oct 2024 at 08:47, dragoncoder047 ***@***.***>
wrote:

> I guess I'm looking for friction then...
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kaplayjs/kaplay/discussions/424#discussioncomment-10838327>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAHIYW5WZMCEOI2FV5Z5E7TZZXJQBAVCNFSM6AAAAABPJ557J6VHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTAOBTHAZTENY>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Kind Regards,

Marc Flerackers
Software Engineer

**Comment by mflerackers** [comment]
@dragoncoder047 Can you check the restitution branch?
It implements restitution and friction.
friction is very primitive, it should work on time, not frame, but then you'd need 50 to stop in one frame. While you probably expect 1 to stop immediately.
It should also work with the normal force, since higher gravity, more horizontal surface, gives more friction.

**Reply by dragoncoder047** [reply]
Friction looks good for the limitations that there are. I added a comment to one of the commits since I couldn't get restitution to work.
